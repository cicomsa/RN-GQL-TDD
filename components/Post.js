import React from 'react'
import { StyleSheet, Text, View, TouchableOpacity } from 'react-native'

const Post = ({ post, deletePost, id }) => {
  console.log(post)
  return (
    // todo - remove <View> ?
    <View>
      <TouchableOpacity activeOpacity={0.8} onPress={deletePost.bind(this, id)}>
        <View style={styles.post}>
          <Text style={styles.title}>{post.author}</Text>
        </View>
      </TouchableOpacity>
      <View style={styles.post}>
        <Text style={styles.body}>{post.body}</Text>
      </View>
    </View>
  )
}

const styles = StyleSheet.create({
  post: {
    marginVertical: 5,
  },
  title: {
    color: 'purple',
    fontWeight: 'bold',
    textAlign: 'center',
  },
  body: {
    color: 'orange',
    textAlign: 'center'
  }
})

export default Post