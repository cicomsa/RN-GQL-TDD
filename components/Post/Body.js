import React, { useContext } from 'react'
import { StyleSheet, Text, View } from 'react-native'
import { PostContent } from '../Posts'

const Body = () => {
  const { body } = useContext(PostContent)

  return (
    <View style={styles.post}>
      <Text style={styles.body}>{body}</Text>
    </View>
  )
}

const styles = StyleSheet.create({
  post: {
    marginVertical: 5,
  },
  body: {
    color: 'orange',
    textAlign: 'center'
  }
})

export default Body