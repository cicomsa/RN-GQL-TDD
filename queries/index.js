import gql from 'graphql-tag'

const GET_POSTS = gql`
  query posts {
    posts {
      id
      author
      body
    }
  }
`

const ADD_POST = gql`
  mutation addPost($input: PostInput!) {
    addPost(input: $input) {
      id
    }
  }
`


const DELETE_POST = gql`
  mutation deletePost($id: ID!) {
    deletePost(id: $id) {
      body
    }
  }
`

export {
  GET_POSTS,
  ADD_POST,
  DELETE_POST
}
