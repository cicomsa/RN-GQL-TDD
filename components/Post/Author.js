import React, { useContext } from 'react'
import { StyleSheet, Text, View, TouchableOpacity } from 'react-native'
import { GET_POSTS, DELETE_POST } from '../../queries'
import client from '../../apollo-client'
import { PostContent } from '../../contexts'

const Author = () => {
  const { author, id } = useContext(PostContent)
  const deletePost = async id => {
    await client.mutate({
      variables: { id },
      mutation: DELETE_POST,
      refetchQueries: () => [{ query: GET_POSTS }]
    })
  }

  return (
    <TouchableOpacity className="author" activeOpacity={0.8} onPress={() => deletePost(id)}>
      <Text className="author__copy" style={styles.author}>{author}</Text>
    </TouchableOpacity>
  )
}

const styles = StyleSheet.create({
  author: {
    color: 'purple',
    fontWeight: 'bold',
    marginVertical: 5,
    textAlign: 'center',
  }
})

export default Author