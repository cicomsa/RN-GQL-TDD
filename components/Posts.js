import React, { useState } from 'react'
import { StyleSheet, Text, View, FlatList } from 'react-native'
import { Query } from "react-apollo"
import client from '../apollo-client'
import { GET_POSTS } from '../queries'
import Post from './Post'
import { PostContent } from '../contexts'

const Posts = () => {
  return (
    <Query
      query={GET_POSTS}
    >
      {({ loading, error, data }) => {

        if (loading) return <View><Text>Loading...</Text></View>;
        if (error) return <View><Text>Error :(</Text></View>;

        return (
          <FlatList
            data={data.posts}
            keyExtractor={item => item.id}
            renderItem={data =>
              <PostContent.Provider value={{ author: data.item.author, id: data.item.id, body: data.item.body }}>
                <Post />
              </PostContent.Provider>
            }
          />
        )
      }}
    </Query>
  )
}

const styles = StyleSheet.create({
  posts: {
    marginTop: 5
  }
})

export default Posts