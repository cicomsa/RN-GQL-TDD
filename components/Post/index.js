import React from 'react'
import Author from './Author'
import Body from './Body'

const Post = () => {
  return (
    <>
      <Author />
      <Body />
    </>
  )
}

export default Post