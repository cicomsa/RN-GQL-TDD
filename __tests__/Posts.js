import React from 'react'
import { ApolloProvider } from 'react-apollo';
import renderer from 'react-test-renderer'
import { mount, shallow } from 'enzyme'
import Posts from '../components/Posts'
import { GET_POSTS } from '../queries'
import { MockedProvider } from '@apollo/react-testing';
import { Text } from 'react-native'
import wait from 'waait'
import Post from '../components/Post'

const mocks = [
  {
    request: {
      query: GET_POSTS
    },
    result: {
      data: {
        posts: [
          { id: '1', author: 'Author1', body: 'Body1' },
          { id: '2', author: 'Author2', body: 'Body2' }
        ]
      },
    }
  },
];

const mocksError = [
  {
    request: {
      query: GET_POSTS
    },
    error: new Error('error')
  },
];

describe('Posts', () => {
  it('should render loading state correctly', () => {
    const tree = renderer.create(
      <MockedProvider mocks={mocks} addTypename={false}>
        <Posts />
      </MockedProvider>
    ).toJSON()
    expect(tree).toMatchSnapshot()
  })
  it('should render loading state initially', () => {
    const component = mount(
      <MockedProvider mocks={[]}>
        <Posts />
      </MockedProvider>
    );
    expect(component.find(Text).exists()).toBe(true)
    expect(component.find(Text).at(0).text()).toBe('Loading...')
  });
  it('should render two Post components', async () => {
    const wrapper = renderer.create(
      <MockedProvider mocks={mocks} addTypename={false}>
        <Posts />
      </MockedProvider>
    )

    await wait(0);

    const component = wrapper.root.findAllByType(Post);
    expect(component.length).toBe(2);

    const tree = wrapper.toJSON()
    expect(tree).toMatchSnapshot()
  });
  it('should render error message', async () => {
    const wrapper = renderer.create(
      <MockedProvider mocks={mocksError} addTypename={false}>
        <Posts />
      </MockedProvider>
    )
    await wait(0); // wait for response

    const component = wrapper.root.findByType(Text);
    expect(component.props.children).toEqual('Error :(')

    const tree = wrapper.toJSON()
    expect(tree).toMatchSnapshot()
  });
})

