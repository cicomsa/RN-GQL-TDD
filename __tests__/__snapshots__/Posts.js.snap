// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Posts should render error message 1`] = `
<View>
  <Text>
    Error :(
  </Text>
</View>
`;

exports[`Posts should render loading state correctly 1`] = `
<View>
  <Text>
    Loading...
  </Text>
</View>
`;

exports[`Posts should render two Post components 1`] = `
<RCTScrollView
  data={
    Array [
      Object {
        "author": "Author1",
        "body": "Body1",
        "id": "1",
      },
      Object {
        "author": "Author2",
        "body": "Body2",
        "id": "2",
      },
    ]
  }
  disableVirtualization={false}
  getItem={[Function]}
  getItemCount={[Function]}
  horizontal={false}
  initialNumToRender={10}
  keyExtractor={[Function]}
  maxToRenderPerBatch={10}
  numColumns={1}
  onContentSizeChange={[Function]}
  onEndReachedThreshold={2}
  onLayout={[Function]}
  onMomentumScrollEnd={[Function]}
  onScroll={[Function]}
  onScrollBeginDrag={[Function]}
  onScrollEndDrag={[Function]}
  removeClippedSubviews={false}
  renderItem={[Function]}
  scrollEventThrottle={50}
  stickyHeaderIndices={Array []}
  updateCellsBatchingPeriod={50}
  viewabilityConfigCallbackPairs={Array []}
  windowSize={21}
>
  <View>
    <View
      onLayout={[Function]}
      style={null}
    >
      <View
        accessible={true}
        focusable={true}
        isTVSelectable={true}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Object {
            "opacity": 1,
          }
        }
      >
        <Text
          className="author__copy"
          style={
            Object {
              "color": "purple",
              "fontWeight": "bold",
              "marginVertical": 5,
              "textAlign": "center",
            }
          }
        >
          Author1
        </Text>
      </View>
      <View
        style={
          Object {
            "marginVertical": 5,
          }
        }
      >
        <Text
          style={
            Object {
              "color": "orange",
              "textAlign": "center",
            }
          }
        >
          Body1
        </Text>
      </View>
    </View>
    <View
      onLayout={[Function]}
      style={null}
    >
      <View
        accessible={true}
        focusable={true}
        isTVSelectable={true}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Object {
            "opacity": 1,
          }
        }
      >
        <Text
          className="author__copy"
          style={
            Object {
              "color": "purple",
              "fontWeight": "bold",
              "marginVertical": 5,
              "textAlign": "center",
            }
          }
        >
          Author2
        </Text>
      </View>
      <View
        style={
          Object {
            "marginVertical": 5,
          }
        }
      >
        <Text
          style={
            Object {
              "color": "orange",
              "textAlign": "center",
            }
          }
        >
          Body2
        </Text>
      </View>
    </View>
  </View>
</RCTScrollView>
`;
